util/cosmos.js

```diff
@@ -0,0 +1,290 @@
+const config = require('../config/config.js');
+const {
+  db,
+  txCollection: txLogRef,
+} = require('./firebase');
+const publisher = require('./gcloudPub');
+
+const PUBSUB_TOPIC_MISC = 'misc';
+const {
+  COSMOS_LCD_ENDPOINT,
+  COSMOS_BLOCK_TIME = 5000,
+  COSMOS_GAS = '200000',
+  COSMOS_DENOM = 'nanolike',
+  COSMOS_CHAIN_ID = '',
+} = config;
+
+const api = axios.create({
+  baseURL: `http://${COSMOS_LCD_ENDPOINT}`,
+  httpAgent: new http.Agent({ keepAlive: true }),
+  httpsAgent: new https.Agent({ keepAlive: true }),
+});
+
+function timeout(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+let currentHeightNumber;
+
+async function getCurrentHeight() {
+  const res = await api.get('/blocks/latest');
+  const { block_meta: { header: height } } = res.data;
+  return height;
+}
+
+async function updateCurrentHeight() {
+  try {
+    currentHeightNumber = await getCurrentHeight();
+  } catch (err) {
+    console.error(err);
+  }
+  setTimeout(() => updateCurrentHeight(), COSMOS_BLOCK_TIME);
+}
+updateCurrentHeight();
+
+const STATUS = {
+  // PENDING is the initial status of the transaction in database
+  PENDING: 'pending',
+
+  // SUCCESS, FAIL, TIMEOUT status will be written into database
+  SUCCESS: 'success',
+  FAIL: 'fail',
+  TIMEOUT: 'timeout',
+
+  // NOT_FOUND, MINED, CONFIRMED status will be used in this app internally only
+  NOT_FOUND: 'not found',
+  MINED: 'mined',
+  CONFIRMED: 'confirmed',
+};
+
+function createSigner(privKey) {
+  const privateKey = Buffer.from(privKey, 'hex');
+  const publicKey = secp256k1.publicKeyCreate(privateKey, true);
+  const sha256 = crypto.createHash('sha256');
+  const ripemd = crypto.createHash('ripemd160');
+  sha256.update(publicKey);
+  ripemd.update(sha256.digest());
+  const rawAddr = ripemd.digest();
+  const cosmosAddress = bech32.encode('cosmos', bech32.toWords(rawAddr));
+  const signer = (msg) => {
+    const msgSha256 = crypto.createHash('sha256');
+    msgSha256.update(Buffer.from(msg, 'utf-8'));
+    const msgHash = msgSha256.digest();
+    const { signature } = secp256k1.sign(msgHash, privateKey);
+    return { signature, publicKey };
+  };
+  return {
+    cosmosAddress,
+    signer,
+  };
+}
+
+const {
+  cosmosAddress,
+  signer,
+} = createSigner(cosmosKey);
+
+async function getTransactionStatus(txHash) {
+  try {
+    const { data: networkTx } = await api.get(`/txs/${txHash}`);
+    if (!networkTx) {
+      return { status: STATUS.NOT_FOUND };
+    }
+    if (!networkTx.height) {
+      return { status: STATUS.PENDING };
+    }
+    if (!currentHeightNumber) {
+      currentHeightNumber = await getCurrentHeight();
+    }
+    if (networkTx.code && networkTx.code !== '0') {
+      console.error(`${networkTx.code}: ${networkTx.message}`); // eslint-disable-line no-console
+      return { status: STATUS.FAIL };
+    }
+    const { logs: [{ success }] } = networkTx;
```

Comment:
For multiple message, the first messages could success while the last one failed (which fails the whole transaction), so should check the last entry in `logs`